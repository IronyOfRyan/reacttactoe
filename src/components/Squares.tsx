import { SquareCSS } from "../view/styles";
import { calculateWinner } from "../util/gameAI";

export interface SquaresProps {
  state: any;
  handleClick: any;
  i: number; 
}

export interface SquareCSSProps {
  player: string;
  value: boolean;
  onClick?: (() => void);
  onTouchEnd?: (() => void);
  disabled: boolean;
}
// square Component with event listeners
 const Square: React.FunctionComponent<SquaresProps> = ({ state, handleClick, i }) => {
  let winner = calculateWinner(state.squares);
  let squares = [...state.squares];
  
  if (squares[i] === "O") {
    return (
      <SquareStyle disabled={true} player={"player"} value={state.squares[i]} />
    );
    // Does the same for AI
  } else if (squares[i] === "X") {
    return <SquareStyle disabled={true} player={"AI"} value={state.squares[i]} />;
  } else {
    // Default rendered squares when ...squares[i] is empty(nulled)
    return (
      <SquareStyle
        player={''}
        disabled={winner ? true : !state.myTurn}
        value={state.squares[i]}
        onTouchEnd={() => handleClick(i)}
        onClick={() => handleClick(i)}
      />
    );
  }
};

const SquareStyle: React.FunctionComponent<SquareCSSProps> = ({ player, onTouchEnd, onClick, disabled }) => {
  return (
    <SquareCSS
      player={player}
      children
      style={{
        pointerEvents: disabled ? "none" : "auto",
      }}
      onTouchMove={onTouchEnd}
      onClick={onClick}
    />
  );
};

export default Square;
