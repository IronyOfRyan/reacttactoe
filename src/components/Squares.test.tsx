import ReactDOM from "react-dom";
import { render } from '@testing-library/react';
import Squares from "./Squares";

const state = {squares: [], myTurn: false} 
const handleClick = (() => null);

test("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Squares i={0} state={state} handleClick={handleClick}/>, div);
});

test('check default square color', () => {
  const { container } = render(<Squares i={0} state={state} handleClick={handleClick}/>);
  const square = container.querySelector('.sc-iCoGMd.jYXntD')
  expect(square).toHaveStyleRule('background', '#cbb4d4')
  expect(square).toMatchSnapshot()
  expect(square).toBeInTheDocument();
});