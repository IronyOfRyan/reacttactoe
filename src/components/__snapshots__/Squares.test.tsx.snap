// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check default square color 1`] = `
.c0 {
  background: #cbb4d4;
  color: #F00000;
  border: 1px dashed #cbb4d4;
  border-radius: 2px;
  box-shadow: 0 0 0 4px #cbb4d4,2px 1px 6px 4px rgba(10,10,0,0.5);
  font-family: Fascinate,sans-serif;
  width: 20.833333333333336vw;
  min-height: 20.833333333333336vw;
  margin: .6em;
  text-align: center;
  text-shadow: 1px 0.6px #grey;
}

.c0:focus {
  outline: none;
}

.c0:hover {
  background: #5F9EA0;
  border: 1px dashed #5F9EA0;
  border-radius: 1px;
  box-shadow: 0 0 0 4px #5F9EA0,2px 1px 6px 4px rgba(10,10,0,0.5);
  opacity: 1;
}

@media (min-width:768px) {
  .c0 {
    width: 17.36111111111111vw;
    min-height: 17.36111111111111vw;
    margin: .6em;
  }
}

@media (min-width:1024px) {
  .c0 {
    width: 8.333333333333332vw;
    min-height: 8.333333333333332vw;
  }
}

<button
  class="c0"
  style="pointer-events: none;"
/>
`;
